smart_open/gcs.py

```diff
@@ -0,0 +1,523 @@
+    ):
+        if not client:
+            client = google.cloud.storage.Client()
+        bucket = client.get_bucket(bucket)  # type: google.cloud.storage.Bucket
+
+        self._blob = bucket.get_blob(key)
+        if self._blob is None:
+            raise google.cloud.exceptions.NotFound('blob {} not found in {}'.format(key, bucket))
+        self._size = self._blob.size if self._blob.size is not None else 0
+
+        self._raw_reader = SeekableRawReader(self._blob, self._size)
+        self._current_pos = 0
+        self._buffer_size = buffering
+        self._buffer = smart_open.bytebuffer.ByteBuffer(buffering)
+        self._eof = False
+        self._line_terminator = line_terminator
+
+        #
+        # This member is part of the io.BufferedIOBase interface.
+        #
+        self.raw = None
+
+    #
+    # Override some methods from io.IOBase.
+    #
+    def close(self):
+        """Flush and close this stream."""
+        logger.debug("close: called")
+        self._blob = None
+
+    def readable(self):
+        """Return True if the stream can be read from."""
+        return True
+
+    def seekable(self):
+        """If False, seek(), tell() and truncate() will raise IOError.
+
+        We offer only seek support, and no truncate support."""
+        return True
+
+    #
+    # io.BufferedIOBase methods.
+    #
+    def detach(self):
+        """Unsupported."""
+        raise io.UnsupportedOperation
+
+    def seek(self, offset, whence=START):
+        """Seek to the specified position.
+
+        :param int offset: The offset in bytes.
+        :param int whence: Where the offset is from.
+
+        Returns the position after seeking."""
+        logger.debug('seeking to offset: %r whence: %r', offset, whence)
+        if whence not in _WHENCE_CHOICES:
+            raise ValueError('invalid whence, expected one of %r' % _WHENCE_CHOICES)
+
+        if whence == START:
+            new_position = offset
+        elif whence == CURRENT:
+            new_position = self._current_pos + offset
+        else:
+            new_position = self._size + offset
+        new_position = clamp(new_position, 0, self._size)
+        self._current_pos = new_position
+        self._raw_reader.seek(new_position)
+        logger.debug('new_position: %r', self._current_pos)
+
+        self._buffer.empty()
+        self._eof = self._current_pos == self._size
+        return self._current_pos
+
+    def tell(self):
+        """Return the current position within the file."""
+        return self._current_pos
+
+    def truncate(self, size=None):
+        """Unsupported."""
+        raise io.UnsupportedOperation
+
+    def read(self, size=-1):
+        """Read up to size bytes from the object and return them."""
+        if size == 0:
+            return b''
+        elif size < 0:
+            from_buf = self._read_from_buffer()
+            self._current_pos = self._size
+            return from_buf + self._raw_reader.read()
```

Comment:
You don't need to declare variables here.

```suggestion
            return self._read_from_buffer() + self._raw_reader.read()
```